### PR TITLE
Parse and lightly validate Android strings

### PR DESCRIPTION
**Pull request by stevelandeyasana**
This PR officially makes Locheck a cross-platform project!

Adds support for an `androidstrings` command which can parse `strings.xml` and compare two translations to make sure all keys are present in both. Additional validation will follow in another PR but will be slightly simpler than Stringsdict validation.

I tested this by running it on Asana's English and German translations and found a bunch of missing keys that are really missing from German.

### CODE REVIEW

File: Sources/LocheckLogic/Expressions.swift [hunk 1]
```diff
@@ -69,12 +69,19 @@ struct Expressions {
...
-    private static let stringsdictArgumentExpression = #"%#@(?<name>.*?)@"#
+    private static let stringsdictArgumentExpression = #"%#@(?<name>.+?)@"#
     static let stringsdictArgumentRegex = try! NSRegularExpression(
         pattern: Expressions.stringsdictArgumentExpression,
         options: [])
+
+    private static let phraseArgumentExpression = #"\{(?<name>\w+?)\}"#
+    static let phraseArgumentRegex = try! NSRegularExpression(
+        pattern: Expressions.phraseArgumentExpression,
+        options: [])
```
Comment (stevelandeyasana) [inline Sources/LocheckLogic/Expressions.swift:86]: We use a mix of Phrase (`{num_apples} apples`) and "native" (`%d apples`) so I decided to just use different expressions and do some basic auto-detection of format.

File: Sources/LocheckLogic/Types/FormatString.swift [hunk 2]
```diff
@@ -0,0 +1,82 @@
+//
+//  FormatString.swift
```
Comment (stevelandeyasana) [inline Sources/LocheckLogic/Types/FormatString.swift:2]: This was `LocalizedString`.

**Comment by stevelandeyasana** [comment]
Added demo XML files and an integration test. Further work will be on a new branch.

File: Examples/strings-translation.xml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    <!-- Phrase plurals -->
+    <plurals name="plural_number_of_projects">
+        <item quantity="one">{number_of_projects} Projekt</item>
+        <item quantity="other">{number_of_projects} Projekte</item>
+    </plurals>
+
+    <!-- Native plurals -->
+    <plurals name="number_of_errors">
+        <item quantity="one">Ein Fehler ist aufgetreten.</item>
+        <item quantity="other"> %d Fehler sind aufgetreten.</item>
```
Comment (TheDom) [inline Examples/strings-translation.xml:12]: [nit] Remove space before `%d`

**Review by TheDom — APPROVED** [review]
LGTM, that was quick!

File: Sources/LocheckLogic/Types/AndroidStringsFile.swift [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+            guard !(element.attributes["translatable"] == "false") else {
+                continue // skip on purpose!
+            }
+
+            switch element.name {
+            case "string":
+                strings[key] = AndroidString(
+                    key: key,
+                    value: FormatString(string: element.text ?? "", path: path, line: nil))
```
[the remark this answers is not in the document]
Comment (stevelandeyasana) [reply Sources/LocheckLogic/Types/AndroidStringsFile.swift:64]: I plan to make line numbers non-optional again after this. I wrote a patch for SwiftyXMLParser.

File: Sources/LocheckLogic/Expressions.swift [hunk 5]
```diff
@@ -69,12 +69,19 @@ struct Expressions {
 
     // Technically length modifiers are invalid for @ and potentially some others, but in practice
     // it probably doesn't matter.
-    private static let argumentExpression =
+    private static let nativeArgumentExpression =
         "%((?<position>\\d+)\\$)?(?<specifier>(\(lengthExpression))?(\(specifierExpression)))"
-    static let argumentRegex = try! NSRegularExpression(pattern: Expressions.argumentExpression, options: [])
+    static let nativeArgumentRegex = try! NSRegularExpression(
```
[the remark this answers is not in the document]
Comment (stevelandeyasana) [reply Sources/LocheckLogic/Expressions.swift:74]: Indeed they do! I think it comes from C.

File: Sources/LocheckLogic/Types/FormatString.swift [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    let kind: Kind
+
+    init(string: String, path: String, line: Int?) {
+        self.string = string
+        self.path = path
+        self.line = line
+
+        let nativeArguments = parseNativeArguments(string: string)
+        arguments = nativeArguments
+        if nativeArguments.isEmpty {
```
[the remark this answers is not in the document]
Comment (stevelandeyasana) [reply Sources/LocheckLogic/Types/FormatString.swift:34]: Will do!

**Pull request merged**